File: skills_manager/adapters.py
"""Client-specific skill directory adapters."""
from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path

from . import store

CLIENTS = ("claude", "codex")


@dataclass(frozen=True)
class ClientAdapter:
    client: str

    def global_dir(self) -> Path:
        if self.client == "claude":
            return store.home() / ".claude" / "skills"
        if self.client == "codex":
            codex_home = os.environ.get("CODEX_HOME")
            base = Path(codex_home).expanduser() if codex_home else store.home() / ".codex"
            return base / "skills"
        raise ValueError(f"unknown client: {self.client}")

    def project_dir(self, project: str | Path) -> Path:
        base = Path(project).expanduser().resolve()
        if self.client == "claude":
            return base / ".claude" / "skills"
        if self.client == "codex":
            return base / ".codex" / "skills"
        raise ValueError(f"unknown client: {self.client}")

    def rendered_dir(self, project: str | Path | None = None) -> Path:
        return self.project_dir(project) if project else self.global_dir()


def expand_clients(client: str) -> list[str]:
    if client == "all":
        return list(CLIENTS)
    if client not in CLIENTS:
        raise ValueError(f"unknown client: {client}")
    return [client]


def adapter(client: str) -> ClientAdapter:
    return ClientAdapter(client)

File: src/core/adapters.ts
import { join, resolve } from "node:path";
import { expandUser, skillsManagerHome, type Env } from "./paths.js";

export const CLIENTS = ["claude", "codex"] as const;
export type ClientName = (typeof CLIENTS)[number];
export type ClientMode = ClientName | "all";

export class ClientAdapter {
  readonly client: ClientName;
  readonly env: Env;

  constructor(client: ClientName, env: Env = process.env) {
    this.client = client;
    this.env = env;
  }

  globalDir(): string {
    if (this.client === "claude") {
      return join(skillsManagerHome(this.env), ".claude", "skills");
    }
    const base = this.env.CODEX_HOME ? resolve(expandUser(this.env.CODEX_HOME, this.env)) : join(skillsManagerHome(this.env), ".codex");
    return join(base, "skills");
  }

  projectDir(project: string): string {
    const base = resolve(expandUser(project, this.env));
    return join(base, `.${this.client}`, "skills");
  }

  renderedDir(project?: string): string {
    return project ? this.projectDir(project) : this.globalDir();
  }
}

export function expandClients(client: ClientMode): ClientName[] {
  if (client === "all") {
    return [...CLIENTS];
  }
  if (!CLIENTS.includes(client)) {
    throw new Error(`unknown client: ${client}`);
  }
  return [client];
}

export function adapter(client: ClientName, env: Env = process.env): ClientAdapter {
  return new ClientAdapter(client, env);
}

